cartography/intel/gcp/spanner.py

```diff
@@ -0,0 +1,417 @@
+import json
+import logging
+import time
+from typing import Dict
+from typing import List
+
+import neo4j
+from googleapiclient.discovery import HttpError
+from googleapiclient.discovery import Resource
+from cloudconsolelink.clouds.gcp import GCPLinker
+
+from . import label
+from cartography.util import run_cleanup_job
+from cartography.util import timeit
+
+logger = logging.getLogger(__name__)
+gcp_console_link = GCPLinker()
+
+
+@timeit
+def get_spanner_instances(spanner: Resource, project_id: str, regions: list, common_job_parameters) -> List[Dict]:
+    instances = []
+    try:
+        req = spanner.projects().instances().list(parent=f"projects/{project_id}")
+        while req is not None:
+            res = req.execute()
+            if res.get('instances'):
+                instances.extend(res.get('instances', []))
+            req = spanner.projects().instances().list_next(previous_request=req, previous_response=res)
+        return instances
+    except HttpError as e:
+        err = json.loads(e.content.decode('utf-8'))['error']
+        if err.get('status', '') == 'PERMISSION_DENIED' or err.get('message', '') == 'Forbidden':
+            logger.warning(
+                (
+                    'Could not retrieve spanner instances on project %s due to permissions issues. Code: %s, Message: %s'
+                ), project_id, err['code'], err['message'],
+            )
+            return []
+        else:
+            raise
+
+
+@timeit
+def transform_instances(instances: List[Dict], project_id: str) -> List[Dict]:
+    transformed_instances = []
+    for instance in instances:
+        instance['project_id'] = project_id
+        instance['region'] = 'global'
+        instance['instance_name'] = instance['name'].split('/')[-1]
+        instance['consolelink'] = ''  # TODO
+        transformed_instances.append(instance)
+    return transformed_instances
+
+
+@timeit
+def load_spanner_instances(session: neo4j.Session, data_list: List[Dict], project_id: str, update_tag: int) -> None:
+    session.write_transaction(load_spanner_instances_tx, data_list, project_id, update_tag)
+
+
+@timeit
+def load_spanner_instances_tx(tx: neo4j.Transaction, data: List[Dict], project_id: str, gcp_update_tag: int) -> None:
+    query = """
+    UNWIND $Records as record
+    MERGE (instance:GCPSpannerInstance{name:record.name})
+    ON CREATE SET
+        instance.firstseen = timestamp()
+    SET
+        instance.lastupdated = $gcp_update_tag,
+        instance.region = record.region,
+        instance.name = record.name,
```

Comment:
name field of api have format `projects/<project>/instances/[a-z][-a-z0-9]*[a-z0-9]`. we load name field as a name of instance only